tools/wptrunner/wptrunner/executors/protocol.py

```diff
@@ -356,14 +362,46 @@ def switch_to_window(self, wptrunner_id):
         """Switch to a window given a wptrunner window id
 
         :param str wptrunner_id: window id"""
-        pass
-
-    def switch_to_frame(self, index):
+        if wptrunner_id is None:
+            return
+
+        stack = [str(item) for item in self.parent.base.window_handles()]
+        while stack:
+            item = stack.pop()
+            if item is None:
+                self._switch_to_parent_frame()
+                continue
+            elif isinstance(item, str):
+                self.parent.base.set_window(item)
+            else:
+                self._switch_to_frame(item)
+
+            try:
+                handle_window_id = self.parent.base.execute_script("return window.__wptrunner_id")
+                if str(handle_window_id) == wptrunner_id:
+                    return
+            except Exception:
+                pass
+            frame_count = self.parent.base.execute_script("return window.length")
+            # None here makes us switch back to the parent after we've processed all the subframes
+            stack.append(None)
+            if frame_count:
+                stack.extend(reversed(range(0, frame_count)))
```

Comment:
Why reverse it?

Reply:
Just so we end up going first to last when we pop from the stack. It probably doesn't matter, but does seem cleaner.

Reply:
Sure, works for me :).